language_service/src/code_action.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use miette::Diagnostic;
+use qsc::{
+    compile::ErrorKind,
+    error::WithSource,
+    line_column::{Encoding, Range},
+};
+use qsc_linter::{AstLint, HirLint};
+
+use crate::{
+    compilation::Compilation,
+    protocol::{CodeAction, CodeActionKind, TextEdit, WorkspaceEdit},
+};
+
+pub(crate) fn get_code_actions(
+    compilation: &Compilation,
+    source_name: &str,
+    range: &Range,
+    encoding: Encoding,
+) -> Vec<CodeAction> {
+    // Compute quick_fixes and other code_actions, and then merge them together
+    quick_fixes(compilation, source_name, range, encoding)
+}
+
+fn quick_fixes(
+    compilation: &Compilation,
+    source_name: &str,
+    range: &Range,
+    encoding: Encoding,
+) -> Vec<CodeAction> {
+    let mut code_actions = Vec::new();
+
+    // get relevant diagnostics
+    let diagnostics = compilation
+        .errors
+        .iter()
+        .filter(|error| is_error_relevant(error, source_name, range, encoding));
+
+    // An example of what quickfixes could look like if they were generated here.
+    // The other option I considered was generating the quickfixes when the errors
+    // are initially issued. But that has two problems:
+    //  1. It does unnecesary computations at compile time, that would go to waste if using the CLI compiler.
+    //  2. The quickfix logic would be spread across many crates in the compiler.
```

Comment:
Yep, I think this approach works best. If we ever hit perf issues as the number of available code actions grows, VS Code provides an opportunity to defer the computation of edits even further (using the `resolveCodeAction` request). But I don't see this being an issue anytime soon.